pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,667 @@
+							continue
+						}
+
+						for _, blob := range srcManifest.References() {
+							// tagging within the same registry is a no-op
+							if src.ref.Registry == dst.ref.Registry && canonicalFrom.String() == canonicalTo.String() {
+								continue
+							}
+
+							var options []distribution.BlobCreateOption
+							blobSource, err := reference.WithDigest(canonicalFrom, blob.Digest)
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unexpected error building named digest: %v", err)})
+								continue
+							}
+							if !o.SkipMount {
+								options = append(options, client.WithMountFrom(blobSource), WithDescriptor(blob))
+							}
+
+							w, err := toBlobs.Create(ctx, options...)
+							if err == ErrAlreadyExists {
+								glog.V(5).Infof("Blob already exists %#v", blob)
+								continue
+							}
+							if ebm, ok := err.(distribution.ErrBlobMounted); ok {
+								glog.V(5).Infof("Blob mounted %#v", blob)
+								if ebm.From.Digest() != blob.Digest {
+									digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to push %s: tried to mount blob %s src source and got back a different digest %s", src.ref, blob.Digest, ebm.From.Digest())})
+									continue
+								}
+								break
+							}
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to upload blob %s to %s: %v", blob.Digest, dst.ref, err)})
+								break
+							}
+							err = func() error {
+								glog.V(5).Infof("Uploading blob %s", blob.Digest)
+								defer w.Cancel(ctx)
+								r, err := srcBlobs.Open(ctx, blob.Digest)
+								if err != nil {
+									return fmt.Errorf("unable to open source layer %s to copy to %s: %v", blob.Digest, dst.ref, err)
+								}
+								defer r.Close()
+
+								switch dst.t {
+								case DestinationS3:
+									fmt.Fprintf(o.ErrOut, "uploading: %s %d s3://%s\n", blob.Digest, blob.Size, dst.ref)
+								default:
+									fmt.Fprintf(o.ErrOut, "uploading: %s %d %s\n", blob.Digest, blob.Size, dst.ref)
+								}
+
+								n, err := w.ReadFrom(r)
+								if err != nil {
+									return fmt.Errorf("unable to copy layer %s to %s: %v", blob.Digest, dst.ref, err)
+								}
+								if n != blob.Size {
+									fmt.Fprintf(o.ErrOut, "warning: Layer size mismatch for %s: had %d, wrote %d\n", blob.Digest, blob.Size, n)
+								}
+								_, err = w.Commit(ctx, blob)
+								return err
+							}()
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: err})
+								break
+							}
+						}
+					}
+				}
+
+				if len(digestErrs) > 0 {
+					continue
+				}
+
+				// upload and tag the manifest
+				for _, tag := range dst.tags {
+					toDigest, err := toManifests.Put(ctx, srcManifest, distribution.WithTag(tag))
```

Comment:
How so?  srcManifest in this case is one of the manifests out of the list, not the whole list.  We push the final manifest last, so it's overwritten.  Agree it could be optimized, will look at that.

Reply:
srcManifest in this case is result of

    srcManifest, err := manifests.Get(ctx, digest.Digest(srcDigest), schema2ManifestOnly)

But it looks like a bug, because manifests from

    for _, srcManifest := range srcManifests

are not created at all.

Reply:
Filter now copies the manifest list to the end if necessary.

Reply:
That won't help. Manifest must not include references that don't exist in the destination repository. Otherwise, [verification](https://github.com/docker/distribution/blob/bb49a1685d2773cb43dacc16e100419f310ba347/registry/storage/schema2manifesthandler.go#L109) will fail.